File: meWrap/Views/StreamView/StreamReusableView.m
//
//  StreamCell.m
//  Moji
//
//  Created by Sergey Maximenko on 8/18/15.
//  Copyright (c) 2015 Ravenpod. All rights reserved.
//

#import "StreamReusableView.h"

@interface StreamReusableView () <UIGestureRecognizerDelegate>

@property (weak, nonatomic) UITapGestureRecognizer *selectTapGestureRecognizer;

@end

@implementation StreamReusableView

- (UIView *)contentView {
    if (_contentView == nil) {
        return self;
    } else {
        return _contentView;
    }
}

- (void)layoutWithMetrics:(StreamMetrics *)metrics {}

- (void)loadedWithMetrics:(StreamMetrics *)metrics {
    UITapGestureRecognizer *gestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(select)];
    gestureRecognizer.delegate = self;
    [self addGestureRecognizer:gestureRecognizer];
    self.selectTapGestureRecognizer = gestureRecognizer;
}

- (void)setEntry:(id)entry {
    _entry = entry;
    [self setup:entry];
}

- (void)setup:(id)entry {}

- (void)resetup {
    [self setup:self.entry];
}

- (void)select:(id)entry {
    [self.metrics select:self.item entry:entry];
}

- (IBAction)select {
    [self select:self.entry];
}

- (void)didDequeue {}

- (void)willEnqueue {}

// MARK: - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    return gestureRecognizer != self.selectTapGestureRecognizer || self.metrics.selectable;
}

@end

File: meWrap/Views/StreamView/StreamReusableView.swift
//
//  StreamReusableView.swift
//  meWrap
//
//  Created by Yura Granchenko on 08/02/16.
//  Copyright © 2016 Ravenpod. All rights reserved.
//

import Foundation

class StreamReusableView: UIView, UIGestureRecognizerDelegate {
    
    func setEntry(entry: AnyObject?) {}
    func getEntry() -> AnyObject? { return nil }
    
    var metrics: StreamMetricsProtocol?
    var item: StreamItem?
    var selected: Bool = false
    let selectTapGestureRecognizer = UITapGestureRecognizer()
    
    func layoutWithMetrics(metrics: StreamMetricsProtocol) {}
    
    func didLoad() {
        selectTapGestureRecognizer.addTarget(self, action: #selector(self.selectAction))
        selectTapGestureRecognizer.delegate = self
        self.addGestureRecognizer(selectTapGestureRecognizer)
    }
    
    @IBAction func selectAction() {
        metrics?.select(self)
    }
    
    func didDequeue() {}
    
    func willEnqueue() {}
    
    func resetup() {}
    
    // MARK: - UIGestureRecognizerDelegate
    
    override func gestureRecognizerShouldBegin(gestureRecognizer: UIGestureRecognizer) -> Bool {
        return gestureRecognizer != selectTapGestureRecognizer || metrics?.selectable ?? false
    }
}

class EntryStreamReusableView<T: AnyObject>: StreamReusableView {
    
    init() {
        super.init(frame: CGRect.zero)
    }
    
    override func setEntry(entry: AnyObject?) {
        self.entry = entry as? T
    }

    override func getEntry() -> AnyObject? {
        return entry
    }
    
    var entry: T? {
        didSet {
            resetup()
        }
    }
    
    func setup(entry: T) {}
    
    func setupEmpty() {}
    
    override func resetup() {
        if let entry = entry {
            setup(entry)
        } else {
            setupEmpty()
        }
    }
}
